expose node ordering to caliper readers

slabasan opened the pull request:
Default to true. Order nodes using caliper's node ordering, which records the order in which functions are called (rather than by alphabetical frame ordering).

Note to review setting `node_ordering=False` if the appropriate metric is not in the caliper file (otherwise, have to explicitly turn off node_ordering when calling the reader.

ilumsden commented:
To-Do's before opening a full PR:
- [ ] Add tests for the overload of the `GraphFrame.filter` overloads.
- [ ] Test special cases in `QueryMatcher` that raise errors.
- [ ] Test new case where an error can be raised by the `GraphFrame.filter` function.
- [ ] Allow queries to end with "*" or "+" wildcards.

Possible future query language features:
* Allow multi-query searches.
* Sub-tree searches.
  * Might end up being supported indirectly with multi-query searches.

ilumsden commented:
@slabasan this PR is ready for review.

ilumsden commented:
Here's a screenshot of an example output using the query language.

![example](https://user-images.githubusercontent.com/28907237/84529577-aecac500-acaf-11ea-98d2-9aa67467590b.png)

ilumsden commented:
Here's the corresponding code. Sorry about it being a TXT file, but GitHub apparently doesn't support any code files as attachments.

[unit_test.txt](https://github.com/LLNL/hatchet/files/4772288/unit_test.txt)

ilumsden commented:
Here's an updated version of the above code and output that shows the output of `GraphFrame.tree` with `metric="time (inc)"`.
![example](https://user-images.githubusercontent.com/28907237/84532375-26025800-acb4-11ea-9474-c699aedaf0d1.png)
[unit_test.txt](https://github.com/LLNL/hatchet/files/4772443/unit_test.txt)

ilumsden commented:
I also added a Jupyter notebook with basic example query of MPI and Infiniband Verbs calls. I placed the notebook in the `hatchet-tutorials` directory in the ASDE shared workspace on LC systems.

ilumsden commented:
This PR should probably wait for #176 to be merged. The bug fixed in that PR can cause some surprisingly large issues with the query language.

ilumsden commented:
After a discussion with @slabasan and @bhatele, I've added a new `squash` keyword argument to the `GraphFrame.filter` function. Right now, it simply calls the `squash` function automatically from within the `filter` function so that users don't have to call it themselves. It was added because, in the future, there might be different "squash" behavior for Pandas-based filters and query language filters. By default, this argument is set to False to preserve backwards compatibility.

ilumsden commented:
#176 has been merged. I've merged the updated `develop` branch back into this branch.

slabasan commented on hatchet/__init__.py:
```diff
@@ -7,6 +7,7 @@
 # flake8: noqa: F401
 
 from .graphframe import GraphFrame
+from .query_matcher import QueryMatcher
```
Do we need to import the `QueryMatcher` here?

ilumsden replied:
We don't need to, but it would be useful if someone was using the low-level API for the query language. Otherwise, they'd need to know about the `query_matcher` module.

slabasan replied:
Ah ok that makes sense.

slabasan commented on hatchet/graphframe.py:
```diff
@@ -274,21 +275,43 @@ def drop_index_levels(self, function=np.mean):
         self.dataframe = agg_df
 
-    def filter(self, filter_function):
+    def filter(self, filter_obj, squash=False):
```
Could you please update the docstrings here with the new parameters? We probably should have documented it beforehand.

ilumsden replied:
Yep. I'll add that.

ilumsden replied:
Just updated the docstring

slabasan commented on hatchet/graphframe.py:
```diff
@@ -274,21 +275,43 @@ def drop_index_levels(self, function=np.mean):
+            filtered_df = dataframe_copy.loc[dataframe_copy["node"].isin(match_set)]
+        else:
+            raise InvalidFilter(
+                "The arugment passed to filter must be a callable, a query path list, or a QueryMatcher object."
```
typo here: arugment

ilumsden replied:
Fixed

slabasan commented on hatchet/query_matcher.py:
```diff
@@ -0,0 +1,442 @@
+        return self
+
+    def rel(self, wildcard_spec=".", filter_func=lambda row: True):
+        """Add another edge and node to the query.
```
Insert new line before `Arguments`

ilumsden replied:
Fixed

slabasan commented on hatchet/tests/query.py:
```diff
@@ -0,0 +1,432 @@
+            node.children[0],
+            node.children[0].children[0],
+            node.children[0].children[0].children[0],
+            # node.children[0].children[0].children[0].children[1],
```
a couple legacy comments here that can be removed

ilumsden replied:
Fixed

ilumsden commented:
There's a new example Jupyter notebook in the tutorial materials in the ASDE shared workspace. This one has both basic examples using `mock_graph_literal` from the `tests` directory and practical examples using the MVAPICH-based kripke data from the SC19 paper data.

ilumsden commented:
The `query-lang` branch has been rebased into two commits:
1. Adds the initial implementation of the query language.
2. Adds the `squash` argument to `GraphFrame.filter` and fixes the behavior of the query language when a GraphFrame's index levels aren't dropped.

slabasan commented on hatchet/graphframe.py:
```diff
@@ -274,21 +275,48 @@ def drop_index_levels(self, function=np.mean):
+        """Filter the dataframe using a user-supplied function.
+
+        Arguments:
+            filter_obj (callable, list, or QueryMatcher): the filter to apply to the GraphFrame. If it is a callable, the filter is applied to the DataFrame. If it is a list or a QueryMatcher, the filter is applied to the Graph.
```
shorten this docstring

slabasan commented on hatchet/query_matcher.py:
```diff
@@ -0,0 +1,443 @@
+
+            def _convert_dict_to_filter(attr_filter):
+                """Converts high-level API attribute filter to a lambda"""
+                compops = ("<", ">", "==", ">=", "<=", "<>", "!=")  # ,
```
maybe a trailing comma?

ilumsden replied:
That was intentional. I was planning on adding the rest of the Python comparison operators to that list later. The next line down is a commented out version of the remainder of the list. I was going to leave it there so that I didn't have to hunt them down later. The comma is part of that comment.

slabasan commented on hatchet/query_matcher.py:
```diff
@@ -0,0 +1,443 @@
+    def match(self, wildcard_spec=".", filter_func=lambda row: True):
+        """Start a query with a root node described by the arguments.
+
+        Arugments:
```
typo here

slabasan commented on hatchet/query_matcher.py:
```diff
@@ -0,0 +1,443 @@
+    def _match_pattern(self, gf, pattern_root):
+        """Try to match the query pattern starting at the provided root node.
+
+        Arugments:
```
typo here

ilumsden commented:
@slabasan the recent push should have addressed all the typos. I've already squashed the changes back into the original two commits.

ilumsden commented:
Just rebased this branch on LLNL/hatchet:develop.

slabasan committed:
expose node ordering to caliper readers

Default to true. Order nodes using caliper's node ordering, which records the
order in which functions are called (rather than by alphabetical frame
ordering).

slabasan committed:
add error check, disable overwriting of node ordering

- add error check if want node ordering but file doesn't support it
- when creating the graph, don't overwrite the node ordering boolean
based on if caliper file has specific metric
- update both caliper reader and caliper native reader

slabasan committed:
force node ordering to be false in unit tests
